include/nbl/builtin/hlsl/format/constants.hlsl

```diff
@@ -0,0 +1,50 @@
+
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_HLSL_FORMAT_CONSTANTS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_FORMAT_CONSTANTS_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace format
+{
+
+
+//rgb19e7, our custom 3 channel, shared exponent, floating point format
+#define RGB19E7_MANTISSA_BITS 19
+#define RGB19E7_MANTISSA_MASK 0x7ffff
+#define RGB19E7_EXPONENT_BITS 7
+#define RGB19E7_EXP_BIAS 63
+#define MAX_RGB19E7_EXP (RGB19E7_EXP_BIAS+1)
+
+#define MAX_RGB19E7_MANTISSA_VALUES (0x1<<RGB19E7_MANTISSA_BITS)
+#define MAX_RGB19E7_MANTISSA (MAX_RGB19E7_MANTISSA_VALUES-1)
+#define MAX_RGB19E7 float(MAX_RGB19E7_MANTISSA)/float(MAX_RGB19E7_MANTISSA_VALUES)*exp2(float(MAX_RGB19E7_EXP))
+
+#define RGB19E7_COMPONENT_INDICES int4(0,0,1,1)
+#define RGB19E7_COMPONENT_BITOFFSETS int4(0,RGB19E7_MANTISSA_BITS,(2*RGB19E7_MANTISSA_BITS)&31,(3*RGB19E7_MANTISSA_BITS)&31)
+#define RGB19E7_G_COMPONENT_SPLIT (32-RGB19E7_MANTISSA_BITS)
+
+//rgb18e7s3, out custom 3 channel, shared exponent signed floating point format
+#define RGB18E7S3_MANTISSA_BITS 18
+#define RGB18E7S3_EXPONENT_BITS RGB19E7_EXPONENT_BITS
+#define RGB18E7S3_EXP_BIAS RGB19E7_EXP_BIAS
+#define MAX_RGB18E7S3_EXP (RGB18E7S3_EXP_BIAS+1)
+
+#define MAX_RGB18E7S3_MANTISSA_VALUES (0x1<<RGB18E7S3_MANTISSA_BITS)
+#define MAX_RGB18E7S3_MANTISSA (MAX_RGB18E7S3_MANTISSA_VALUES-1)
+#define MAX_RGB18E7S3 float(MAX_RGB18E7S3_MANTISSA)/float(MAX_RGB18E7S3_MANTISSA_VALUES)*exp2(float(MAX_RGB18E7S3_EXP))
+
+#define RGB18E7S3_COMPONENT_INDICES RGB19E7_COMPONENT_INDICES
+#define RGB18E7S3_COMPONENT_BITOFFSETS int4(0,RGB18E7S3_MANTISSA_BITS,(2*RGB18E7S3_MANTISSA_BITS)&31,(3*RGB18E7S3_MANTISSA_BITS)&31)
+#define RGB18E7S3_G_COMPONENT_SPLIT (32-RGB18E7S3_MANTISSA_BITS)
```

Comment:
I think we have these floating around in triplicate now, C++, GLSL and here, unify!

Also we can define a macro
```cpp
#ifdef __cplusplus
#define NBL_HLSL_CONSTEXPR constexpr
#else
#define NBL_HLSL_CONSTEXPR static const
#endif
```

Reply:
glm might need a contribution where you add `cosntexpr` constructors and mathematical operators to `ivec4`